### PR TITLE
Issue #52: Data Binding on Comics Activity

### PR DESCRIPTION
**Pull request by SalosV**
**Data Binding on Comics Activity** #52 

- Delete comicsUiModel because implemented databinding.
- Create diferent bindingAdapter for use in xml.
- Create Event for use navigation.
- Create extension function for toast.

**Testing Done:**

- Huawei p20 Lite

**How to test:**

- Run the application and click onComic Category
- You can see the loading while the comics have been loaded
- Then you can see the comics in a scrollable list

**Signed-off-by**:

- Santiago Lozada <santiagolozada95@gmail.com>

### CODE REVIEW

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 1]
```diff
@@ -51,35 +42,34 @@ class ComicsActivity : AppCompatActivity() {
...
         adapter = ComicsAdapter(viewModel::onComicClicked)
         recycler.layoutManager = GridLayoutManager(this, GRID_COLUMS)
         recycler.adapter = adapter
 
-        viewModel.model.observe(this, Observer(::updateUi))
+        viewModel.comics.observe(this, Observer(::showComics))
+        viewModel.messageError.observe(this, Observer(::showMessages))
+        viewModel.navegateTo.observe(this, EventObserver { comic ->
+            toast(comic.title)
+        })
```
Comment (jamacado) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: the method onCreate contain a lot of information, create smaller methods

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 2]
```diff
@@ -1,30 +1,26 @@
...
-import com.architect.coders.mu8.utils.makeItVisible
+import com.architect.coders.mu8.utils.toast
+import com.architect.codes.mu8.comics.Comic
 import com.architect.codes.mu8.comics.ComicsUseCaseImpl
 import com.google.android.material.snackbar.Snackbar
 
 class ComicsActivity : AppCompatActivity() {
 
-    private val progress: ProgressBar by lazy { findViewById<ProgressBar>(R.id.comic_activity_progress) }
     private val recycler: RecyclerView by lazy { findViewById<RecyclerView>(R.id.comic_activity_recycler) }
```
Comment (jamacado) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt:24]: Why don't you use the binding?

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 3]
```diff
@@ -51,35 +42,34 @@ class ComicsActivity : AppCompatActivity() {
...
-                recycler.makeItGone()
-                Snackbar.make(
-                    recycler,
-                    getString(R.string.error_internet_message),
-                    Snackbar.LENGTH_LONG
-                ).show()
-            }
-            is NavigateTo -> Toast.makeText(this, uiModel.comic.title, Toast.LENGTH_LONG).show()
-        }
+
```
Comment (jamacado) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: remove space

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 4]
```diff
@@ -34,14 +30,9 @@ class ComicsActivity : AppCompatActivity() {
...
-        setContentView(R.layout.activity_comics)
 
-        toolbar = findViewById(R.id.comic_activity_toolbar)
-        setSupportActionBar(toolbar)
-        supportActionBar?.setDisplayShowTitleEnabled(false)
-
-        toolbarTitle = findViewById(R.id.toolbar_title)
-        toolbarTitle.text = getString(R.string.comics_name)
+        val binding: ActivityComicsBinding =
+            DataBindingUtil.setContentView(this, R.layout.activity_comics)
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: I think this fits in one line

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 5]
```diff
@@ -51,35 +42,45 @@ class ComicsActivity : AppCompatActivity() {
...
         recycler.layoutManager = GridLayoutManager(this, GRID_COLUMS)
         recycler.adapter = adapter
+    }
 
-        viewModel.model.observe(this, Observer(::updateUi))
+    private fun observerViewModel() {
+        viewModel.comics.observe(this, Observer(::showComics))
+        viewModel.messageError.observe(this, Observer(::showMessages))
+        viewModel.navegateTo.observe(this, EventObserver { comic ->
+            toast(comic.title)
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Can you put this and the closing curly braces/parenthesis on the same line?

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 6]
```diff
@@ -51,35 +42,45 @@ class ComicsActivity : AppCompatActivity() {
...
-                    Snackbar.LENGTH_LONG
-                ).show()
-            }
-            is NavigateTo -> Toast.makeText(this, uiModel.comic.title, Toast.LENGTH_LONG).show()
-        }
+    private fun showComics(comics: List<Comic>) {
+        adapter.comics = comics
+    }
+
+    private fun showMessages(message: String) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Rename as show message in singular, you are just receiving one parameter not multiple ones

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsAdapter.kt [hunk 7]
```diff
@@ -20,24 +17,16 @@ class ComicsAdapter(private val listener: (Comic) -> Unit) : RecyclerView.Adapte
     )
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ComicHolder {
-        return ComicHolder(parent.inflate(R.layout.comic_item_list))
+        return  ComicHolder(parent.bindingInflate(R.layout.comic_item_list))
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsAdapter.kt:20]: Why the extra space?

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt [hunk 8]
```diff
@@ -17,26 +18,25 @@ import androidx.lifecycle.ViewModelProviders
 import com.architect.coders.mu8.R
 import com.bumptech.glide.Glide
 
-fun View.makeItGone() {
-    visibility = View.GONE
+fun Context.toast(message: String) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt:22]: If you have an extension function for toast, can you create one for the snack bar?

**Review by recaldev — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 4 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 5 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 4 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

Comment (aliceresponde) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Its not solved, please nove it in single line.

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 5 above)
Comment (aliceresponde) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: please upload the changes

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 9]
```diff
@@ -51,35 +42,45 @@ class ComicsActivity : AppCompatActivity() {
             )
         }
 
+        binding.viewmodel = viewModel
+        binding.lifecycleOwner = this
+
+        configureToolbar()
+
```
Comment (aliceresponde) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: remove space

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt [hunk 10]
```diff
@@ -51,35 +42,45 @@ class ComicsActivity : AppCompatActivity() {
...
+        configureRecycler()
+
```
Comment (aliceresponde) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: remove space

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsViewModel.kt [hunk 11]
```diff
@@ -2,43 +2,41 @@ package com.architect.coders.mu8.comics
...
         }
     }
 
+    private val _comics = MutableLiveData<List<Comic>>()
+    val comics: LiveData<List<Comic>> get() = _comics
+
+    private val _loading = MutableLiveData<Boolean>()
+    val loading: LiveData<Boolean> get() = _loading
+
+    private val _navegateTo = MutableLiveData<Event<Comic>>()
```
Comment (aliceresponde) [inline presentation/src/main/java/com/architect/coders/mu8/comics/ComicsViewModel.kt:33]: Please  check, that Events, is an Observable, then we  dont need the external MutableLiveData

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 4 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 6 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt (hunk 8 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/utils/PresentationExtensions.kt:22]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 10 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 9 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsAdapter.kt (hunk 7 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsAdapter.kt:20]: Done

File: presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt (hunk 5 above)
Comment (SalosV) [reply presentation/src/main/java/com/architect/coders/mu8/comics/ComicsActivity.kt]: Done

**Review by recaldev — APPROVED** [review]

**Review by aliceresponde — APPROVED** [review]
LG

**Pull request merged**